server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -259,16 +282,31 @@ private static List<MoveAllocationCommand> getMoveCommands(AllocationCommands co
     }
 
     private void setCurrentDesiredBalance(DesiredBalance newDesiredBalance) {
-        if (logger.isTraceEnabled()) {
-            var diff = DesiredBalance.hasChanges(currentDesiredBalance, newDesiredBalance)
-                ? "Diff: " + DesiredBalance.humanReadableDiff(currentDesiredBalance, newDesiredBalance)
-                : "No changes";
-            logger.trace("Desired balance updated: {}. {}", newDesiredBalance, diff);
+        // Update current desired balance if the old value has not been changed already by master fail-over
+        final AtomicReference<DesiredBalance> oldDesiredBalance = new AtomicReference<>();
+        final DesiredBalance updatedDesiredBalance = currentDesiredBalanceRef.updateAndGet(current -> {
+            if (current != DesiredBalance.NOT_MASTER) {
+                oldDesiredBalance.set(current);
+                return newDesiredBalance;
+            } else {
+                return current;
+            }
+        });
+
+        if (updatedDesiredBalance == newDesiredBalance) {
+            if (logger.isTraceEnabled()) {
+                var diff = DesiredBalance.hasChanges(updatedDesiredBalance, newDesiredBalance)
+                    ? "Diff: " + DesiredBalance.humanReadableDiff(updatedDesiredBalance, newDesiredBalance)
```

Comment:
Also the two lines above? `updatedDesiredBalance` -> `oldDesiredBalance` ?

Reply:
Yeah. Just noticed that. See a432406f26761bdd4ca413c31774a9c14ed84cb1